Improved implementation of @bound decorator

mbrowne opened the pull request:
@littledan @nicolo-ribaudo I updated the implementation of the `@bound` decorator (kudos to Nicolo for the new approach) to fix an issue with inheritance—it's the issue I described in the rationale document and deleted as part of this PR. But I'm not sure of the proper way to detect private names, since obviously `this[key]` won't work to access those...for now I'm just doing the check the other way around (seeing if it's *not* a public property); I don't know if there's a better way to do this. I guess it wouldn't make a big difference anyhow, but if there's something like `typeof key == "privateName"`, that would be more explicit and maybe more readable.

mbrowne committed:
improved implementation of @bound decorator, and updated rationale document accordingly

nicolo-ribaudo commented:
If `typeof key === "object"` then it is a private name, because keys are normalized (to string/symbol/PrivateName) before calling the decorator.

littledan commented:
I'm not sure if this is the place where we should be maintaining this decorator. How about we move the `@bound` rationale and implementation to a separate repo, and link it from this repository? Then you can develop it freely.

mbrowne commented:
I already did create a separate repo (and npm package) for the `@bound` decorator:
https://github.com/mbrowne/bound-decorator

But given that implementations for the other decorators from the readme are included in METAPROGRAMMING.md, it seems like the `@bound` implementation should be included there too. If it were just a minor tweak, I would say there's no need to bother updating it in this repo since it's just an example after all...but this change significantly mitigates inheritance-related issues.

Also, the rationale document in this repo has the advantage of continuing the running example from https://github.com/tc39/proposal-decorators/blob/master/README.md, whereas the rationale document in my own repo uses React components as the example. It's been a while since I wrote the rationale document and this is the first time I'm updating it. While certainly things could be added to it, I'm not planning to do that and I doubt any further changes will be needed. So I propose leaving it as-is after this PR and then if there's a further need to update it, it can just be deleted and you could link to https://github.com/mbrowne/bound-decorator/blob/master/MOTIVATION.md instead. If you want to go ahead and do that now, I understand, but like I said I do think there's a benefit in continuing the running example from the readme.

I understand where you're coming from and I don't want this to become a significant maintenance burden...it's your call. But I'll go ahead and update the private name check to `typeof key === "object"` in case it can be merged.

mbrowne committed:
updated private name check

littledan commented:
We've been through a few changes already; I'd prefer to move the references after METAPROGRAMMING.md to just point to your repository at this point. I'm just not the biggest expert here, and so I don't want to be attesting for the correctness and appropriate-ness of your decorator.

mbrowne committed:
removed bound decorator rationale document and provided link instead; added link to maintained bound decorator implementation

mbrowne commented:
Ok, I removed the rationale document and provided a link instead. I also updated the implementation with a note that it's not maintained (and a link to my library)—is that sufficient?

CC @ljharb who perhaps could be a reviewer for this final version of the `@bound` decorator (well, final assuming no more big changes to the decorators API itself that would break it).

littledan reviewed (approved):
Thanks for the improvement in the decorator and for being flexible about this issue.
